packages/gatsby/src/cache/cache-fs.ts

```diff
@@ -0,0 +1,268 @@
+      .read(filePath, this.options)
+      .catch(async err => {
+        if (err.code === `ENOENT`) {
+          throw err
+        }
+        //maybe the file is currently written to, lets lock it and read again
+        try {
+          await this._lock(filePath)
+          return await jsonFileStore.read(filePath, this.options)
+        } catch (err2) {
+          throw err2
+        } finally {
+          await this._unlock(filePath)
+        }
+      })
+    if (data.expireTime <= Date.now()) {
+      //cache expired
+      this.del(key).catch(() => 0 /* ignore */)
+      return undefined
+    }
+    if (data.key !== key) {
+      //hash collision
+      return undefined
+    }
+    return data.val
+  } catch (err) {
+    //file does not exist lets return a cache miss
+    if (err.code === `ENOENT`) {
+      return undefined
+    } else {
+      throw err
+    }
+  }
+})
+
+/**
+ * delete entry from cache
+ */
+DiskStore.prototype.del = wrapCallback(async function (key) {
+  const filePath = this._getFilePathByKey(key)
+  try {
+    if (this.options.subdirs) {
+      //check if the folder exists to fail faster
+      const dir = path.dirname(filePath)
+      await promisify(fs.access)(dir, fs.constants.W_OK)
+    }
+
+    await this._lock(filePath)
+    await jsonFileStore.delete(filePath, this.options)
+  } catch (err) {
+    //ignore deleting non existing keys
+    if (err.code !== `ENOENT`) {
+      throw err
+    }
+  } finally {
+    await this._unlock(filePath)
+  }
+})
+
+/**
+ * cleanup cache on disk -> delete all files from the cache
+ */
+DiskStore.prototype.reset = wrapCallback(async function (): Promise<void> {
+  const readdir = promisify(fs.readdir)
+  const stat = promisify(fs.stat)
+  const unlink = promisify(fs.unlink)
+
+  return await deletePath(this.options.path, 2)
+
+  async function deletePath(fileOrDir, maxDeep): Promise<void> {
+    if (maxDeep < 0) {
+      return
+    }
+    const stats = await stat(fileOrDir)
+    if (stats.isDirectory()) {
+      const files = await readdir(fileOrDir)
+      for (let i = 0; i < files.length; i++) {
+        await deletePath(path.join(fileOrDir, files[i]), maxDeep - 1)
+      }
+    } else if (
+      stats.isFile() &&
+      /[/\\]diskstore-[0-9a-fA-F/\\]+(\.json|-\d\.bin)/.test(fileOrDir)
+    ) {
+      //delete the file if it is a diskstore file
+      await unlink(fileOrDir)
+    }
+  }
+})
+
+/**
+ * locks a file so other forks that want to use the same file have to wait
+ * @param {string} filePath
+ * @returns {Promise}
+ * @private
+ */
+DiskStore.prototype._lock = function _lock(filePath): Promise<void> {
+  return new Promise((resolve, reject) =>
+    innerLock(resolve, reject, filePath)
+  ).then(() => {
+    globalGatsbyCacheLock.set(filePath, Date.now())
+  })
+}
+
+function innerLock(resolve, reject, filePath): void {
+  try {
+    let lockTime = globalGatsbyCacheLock.get(filePath) ?? 0
+    if (lockTime > 0 && Date.now() - lockTime > 10 * 60 * 1000) {
```

Comment:
If stale time is supposed to be 10 seconds then it should be (or am I missing something?):

```suggestion
    if (lockTime > 0 && Date.now() - lockTime > 10 * 1000) {
```

Reply:
Haha. Yeah